curly/upload/request.go

```diff
@@ -0,0 +1,53 @@
+package upload
+
+import (
+	"compress/gzip"
+	"fmt"
+	"io"
+	"log"
+	"mime/multipart"
+	"net/http"
+)
+
+type Uploader struct {
+	c         *http.Client
+	uploadURL string
+}
+
+func New(c *http.Client, uploadURL string) *Uploader {
+	return &Uploader{
+		c:         c,
+		uploadURL: uploadURL,
+	}
+}
+
+func (u *Uploader) Upload(r io.Reader, filename string) (*http.Request, error) {
+
+	rBody, wPipe := io.Pipe()
+	writer := multipart.NewWriter(wPipe)
+	go func() {
+		defer wPipe.Close()
+		defer writer.Close()
+
+		gzfilename := fmt.Sprintf("%s.gz", filename)
+		part, err := writer.CreateFormFile("file", gzfilename)
+
+		gzipW := gzip.NewWriter(part)
+		defer gzipW.Close()
+
+		_, err = io.Copy(gzipW, r)
+		if err != nil {
```

Comment:
having log.Fatal here is incorrect because you are not closing connections both from website you are downloading and also to website you are uploading - if they are implemented poorly - this will lead to resource leaks